### ISSUE TITLE
Analyzer for class names (BEM)

### ISSUE THREAD

**Issue by MalteJanz** [MalteJanz/ludtwig#14]
Only check `class` attributes (not `:class` attributes which contain javascript).
Check for:
- `-` as seperator between words
- only zero or one time `__` as seperator
- only zero or one time `--` as seperator

Take a look at http://getbem.com/naming/

**Comment by MalteJanz** [comment]
See #21 why it is currently a bad idea to implement this. The analyzer system needs a rework to be really useful.

**Comment by marcelbrode** [comment]
One thing i'd like to add: check for BEM-prefixes. For Example:
Error level: A component `sw-my-component` should have the following class on the first block: `sw-my-component` 
Warning level: Every further class tag should contain at least one BEM-prefix: `sw-my-component__description`

**Comment by MalteJanz** [comment]
I think i dont fully understand your idea but i try to give you an overview of how ludtwig represents the template in code (the AST - abstract syntax tree).

Ludtwig only sees some kind of tag (it must not be an `div` or something simelar - it can almost have any name without whitespace). So ludtwig doesn't know / care if some tag is a vue component or not. I mean you could check if the tag name contains a `-` to kind of detect a component but that is maybe not the cleanest solution. Also ludtwig doesn't know if the current twig file represents a vue component or even if the twig file name is the component name.

Basically a `Tag` looks like this in the [AST](https://github.com/MalteJanz/ludtwig-parser/blob/f70eeb834005da4f51fc11557c783b748b231a99/src/ast.rs#L256):
```rust
pub struct Tag {
    pub name: String,
    pub self_closed: bool,
    pub attributes: Vec<TagAttribute>,
    pub children: Vec<SyntaxNode>,
}
```
(`Vec` is a list / array with dynamic size)
Where `TagAttribute` is also a more complex structure (it is an enum with values as it variants) to support twig syntax inside tag attributes. But you can iterate over the attributes and look for the basic `HtmlAttribute` which is defined as:
```rust
pub struct HtmlAttribute {
    pub name: String,
    pub value: Option<String>,
}
``` 
(`Option` means that the value is optional)

If you also want to use information of a block it looks like this:
```rust
pub struct TwigBlock<C> {
    pub name: String,
    pub children: Vec<C>,
}
```
(`C` is the type which is variable / generic. This allows it to be used inside of the template and inside of tag attributes)

So to better understand your idea it would be nice to see some kind of short twig template example with the explainations what classes should ludtwig check and how they are defined. In the best case also where the information is coming from 😄.

The idea behind this issue was to simply check if the class name is a valid "BEM" name (has a valid syntax). Everything else is kind of complicated (for example enforcing every class name to start in the same way), because that could limit the developer if it does not always apply.

**Comment by marcelbrode** [comment]
My thought was to check (on error level) that the very first classname should be only the "root" of BEM. 
`sw-my-compontent` is the component name and like that should be the content of the very first class on the first HTML-element.
You just said it'd be difficult to check if classes start like this, so the warnign level type would not be possible. But thats okay, of course :) 

`{% block sw_component %}
    <div class="sw-component">
        {# ... #}
    </div>
{% endblock %}`

Just suggested to check if component-name/filepame === first class.
But I should have a deeper look into rust in the future. Then my understanding will improve :)

**Comment by MalteJanz** [comment]
closed because of #44
